### PR TITLE
Use Album Disambiguation/Type when searching and picking releases

### PR DESCRIPTION
**Pull request by gismo2004**
#### Database Migration
NO

#### Description
Use Album Disambiguation/Type when searching and picking releases

#### Todos
- [ ] Tests
- [ ] Documentation

#### Issues Fixed or Closed by this PR
Use Album Disambiguation/Type when searching and picking releases #432
*

### CODE REVIEW

**Comment by gismo2004** [comment]
I have tested for my indexer, and the result seems to be better than without --> was that the intension @Qstick? If yes, i would add this to other searchgenerators too?

With  Disambiguation:
`18-9-15 13:31:42.7|Trace|HttpClient|Res: [GET] myindexer/api?t=search&cat=3000,3010,3020,3030,3040&extended=1&apikey=(removed)&offset=0&limit=100&q=Weezer%20Weezer%20White%20Album: 200.OK (195 ms)
18-9-15 13:31:42.7|Trace|NewznabRssParser|Parsed: Weezer - Weezer (White Album) (2016) MP3
18-9-15 13:31:42.7|Trace|NewznabRssParser|Parsed: weezerWeezer - Weezer [White Album] (2016) [FLAC 24-96].par2
18-9-15 13:31:42.7|Debug|NzbSearchService|Total of 2 reports were found for [Weezer - Weezer (White Album) (2016)] from 1 indexers`

Without:
`18-9-15 13:33:59.0|Trace|HttpClient|Res: [GET] myindexer/api?t=search&cat=3000,3010,3020,3030,3040&extended=1&apikey=(removed)&offset=0&limit=100&q=Weezer%20Weezer: 200.OK (445 ms)
18-9-15 13:33:59.0|Trace|NewznabRssParser|Parsed: Weezer - Pinkerton (2010 Deluxe Edition Disc 2) (2010) FLAC
18-9-15 13:33:59.1|Trace|NewznabRssParser|Parsed: Weezer - Pinkerton (2010 Deluxe Edition Disc 1) (1996) FLAC
18-9-15 13:33:59.1|Trace|NewznabRssParser|Parsed: Weezer - Weezer (The Blue Album Deluxe Edition Disc 2) (2004) FLAC
18-9-15 13:33:59.1|Trace|NewznabRssParser|Parsed: Weezer - Weezer (The Blue Album Deluxe Edition Disc 1) (1994) FLAC
18-9-15 13:33:59.1|Trace|NewznabRssParser|Parsed: Weezer-Weekend_Woman-SINGLE-WEB-2017-ENTiTLED
18-9-15 13:33:59.1|Trace|NewznabRssParser|Parsed: Weezer-Happy_Hour-SINGLE-WEB-2017-ENTiTLED
18-9-15 13:33:59.1|Trace|NewznabRssParser|Parsed: (Mad) ( Divers W ) "Weezer-Raditude_Deluxe_Edition-Advance-2CD-2009-YOURCANADiANFRiENDS ) ( 192k VBR.zip"
18-9-15 13:33:59.1|Trace|NewznabRssParser|Parsed: (01/13) - Weezer - Pinkerton (1996) FLAC - "Weezer - Pinkerton (1996).par2" - 272,20 MB
18-9-15 13:33:59.1|Trace|NewznabRssParser|Parsed: (Xmas Jamboree) File 82 of 95 - Weezer - The Christmas Song.mp3
18-9-15 13:33:59.1|Trace|NewznabRssParser|Parsed: Weezer - Weezer (White Album) (2016) MP3
18-9-15 13:33:59.1|Trace|NewznabRssParser|Parsed: weezerWeezer - Weezer [White Album] (2016) [FLAC 24-96].par2
18-9-15 13:33:59.2|Trace|NewznabRssParser|Parsed: Weezer-Weezer-CD-FLAC-2016-FORSAKEN
18-9-15 13:33:59.2|Trace|NewznabRssParser|Parsed: Weezer - Discography (AVH)
18-9-15 13:33:59.2|Trace|NewznabRssParser|Parsed: Weezer-Weezer-2016-FNT
18-9-15 13:33:59.2|Trace|NewznabRssParser|Parsed: Weezer-Weezer_(White_Album)-WEB-2016-ESG
18-9-15 13:33:59.2|Trace|NewznabRssParser|Parsed: Weezer-Weezer-2016-FNT
18-9-15 13:33:59.2|Trace|NewznabRssParser|Parsed: Weezer-Weezer-2016-FNT
18-9-15 13:33:59.2|Trace|NewznabRssParser|Parsed: (????) [363/466] - "Weezer - Beverly Hills.mp3"
18-9-15 13:33:59.2|Trace|NewznabRssParser|Parsed: Weezer-California_Kids-WEB-2016-ENTiTLED
18-9-15 13:33:59.2|Trace|NewznabRssParser|Parsed: Weezer-California_Kids-WEB-2016-ENTiTLED
18-9-15 13:33:59.2|Trace|NewznabRssParser|Parsed: (????) [506/567] - "Weezer - Beverly Hills.mp3"
18-9-15 13:33:59.3|Trace|NewznabRssParser|Parsed: (????) [360/461] - "Weezer - Beverly Hills.mp3"
18-9-15 13:33:59.3|Trace|NewznabRssParser|Parsed: (????) [507/568] - "Weezer - Beverly Hills.mp3"
18-9-15 13:33:59.3|Trace|NewznabRssParser|Parsed: Weezer-Thank_God_For_Girls-WEB-2015-SPANK
18-9-15 13:33:59.3|Trace|NewznabRssParser|Parsed: Weezer-Thank_God_For_Girls-WEB-2015-SPANK
18-9-15 13:33:59.3|Trace|NewznabRssParser|Parsed: Weezer-Back_To_The_Shack_(The_Tonight_Show_2014-07-23)-720p-x264-2014-SRP
18-9-15 13:33:59.3|Trace|NewznabRssParser|Parsed: Weezer-Hash_Pipe-(497_559-2)-CDS-FLAC-2001-CUSTODES
18-9-15 13:33:59.3|Trace|NewznabRssParser|Parsed: Mammary Lane Jun 30 2014 Mon [2/6] "Stern-2014_06_30-02-Weezer in studio [2005-10-17].mp3" 17.9MiB
18-9-15 13:33:59.3|Trace|NewznabRssParser|Parsed: Weezer - Make Believe
18-9-15 13:33:59.4|Trace|NewznabRssParser|Parsed: Weezer-Raditude-CD-FLAC-2009-CUSTODES
18-9-15 13:33:59.4|Trace|NewznabRssParser|Parsed: Music Special May 28 2012 Mon [76/77] "Music of Howard Stern - 2012-05-28 - Weezer - Beverly Hills.mp3" 6.5MiB
18-9-15 13:33:59.4|Trace|NewznabRssParser|Parsed: Memorial Day 2011 Music Special [204/208] "Music of Howard Stern - 2011-05-30 - Weezer - Hash Pipe.mp3" 2.5MiB
18-9-15 13:33:59.4|Trace|NewznabRssParser|Parsed: Memorial Day 2011 Music Special [203/208] "Music of Howard Stern - 2011-05-30 - Weezer - Beverly Hills.mp3" 2.7MiB
18-9-15 13:33:59.4|Debug|NzbSearchService|Total of 33 reports were found for [Weezer - Weezer (White Album) (2016)] from 1 indexers`

**Comment by Qstick** [comment]
Yep exactly

File: src/NzbDrone.Core/IndexerSearch/Definitions/AlbumSearchCriteria.cs [hunk 1]
```diff
@@ -7,12 +8,14 @@ public class AlbumSearchCriteria : SearchCriteriaBase
 
         public string AlbumTitle { get; set; }
         public int AlbumYear { get; set; }
+        public string Disambiguation { get; set; }
 
         public string AlbumQuery => GetQueryTitle(AlbumTitle);
+        public string DisambiguationQuery => GetQueryTitle(Disambiguation ?? string.Empty);
```
Comment (Qstick) [inline src/NzbDrone.Core/IndexerSearch/Definitions/AlbumSearchCriteria.cs]: Not sure we need another variable here, should we just include in album query if it’s not empty? Given it’s part of the album.

File: src/NzbDrone.Core/Indexers/Newznab/NewznabRequestGenerator.cs [hunk 2]
```diff
@@ -78,7 +78,8 @@ public virtual IndexerPageableRequestChain GetSearchRequests(AlbumSearchCriteria
                 pageableRequests.AddTier();
 
                 pageableRequests.Add(GetPagedRequests(MaxPages, Settings.Categories, "search",
-                    NewsnabifyTitle($"&q={searchCriteria.ArtistQuery}+{searchCriteria.AlbumQuery}")));
+                    NewsnabifyTitle($"&q={searchCriteria.ArtistQuery}+{searchCriteria.AlbumQuery}" +
+                                    (searchCriteria.DisambiguationQuery.IsNullOrWhiteSpace() ? string.Empty : $"+{searchCriteria.DisambiguationQuery}"))));
```
Comment (Qstick) [inline src/NzbDrone.Core/Indexers/Newznab/NewznabRequestGenerator.cs]: This won’t be needed if we make the other change

File: src/NzbDrone.Core/IndexerSearch/Definitions/AlbumSearchCriteria.cs (hunk 1 above)
Comment (gismo2004) [reply src/NzbDrone.Core/IndexerSearch/Definitions/AlbumSearchCriteria.cs]: yeah, good idea --> my only concern: is this giving good results on "SupportsAudioSearch" indexers? if yes --> much less work :-)

Comment (Qstick) [reply src/NzbDrone.Core/IndexerSearch/Definitions/AlbumSearchCriteria.cs]: Good point, something we should probably test

**Comment by gismo2004** [comment]
i hope someone can test it --> i can't

**Comment by Qstick** [comment]
I’ll test with geek

**Comment by Qstick** [comment]
Works on Geek with artist/album parameter query, not really sure of the others that support artist/album query. Given the very few albums that have disambiguation [7000 out of 1,650,000 in MB database], I'd say we are safe to push it in. (We are better with it than without)

File: src/NzbDrone.Core/IndexerSearch/Definitions/AlbumSearchCriteria.cs [hunk 3]
```diff
@@ -7,12 +8,13 @@ public class AlbumSearchCriteria : SearchCriteriaBase
 
         public string AlbumTitle { get; set; }
         public int AlbumYear { get; set; }
+        public string Disambiguation { get; set; }
 
-        public string AlbumQuery => GetQueryTitle(AlbumTitle);
+        public string AlbumQuery => GetQueryTitle(AlbumTitle+(Disambiguation.IsNullOrWhiteSpace() ? string.Empty : $"+{Disambiguation}"));
```
Comment (Qstick) [inline src/NzbDrone.Core/IndexerSearch/Definitions/AlbumSearchCriteria.cs]: Can you change this to string interpolation like you have the other string.

Something like:
`$"{AlbumTitle}{(Disambiguation.IsNullOrWhiteSpace() ? string.Empty : $"+{Disambiguation}")}"`

Should be good to merge then

**Comment by gismo2004** [comment]
Whats wrong? "Works on my desk"?

**Comment by Qstick** [comment]
It's good, don't mind Travis

**Pull request merged**